src/runtime_src/core/pcie/tools/cloud-daemon/common.cpp

```diff
@@ -144,91 +161,88 @@ int waitForMsg(pcieFunc& dev, int localfd, int remotefd, long interval)
 
-    if (FD_ISSET(localfd, &fds)) {
-        retfd = localfd;
-        dev.log(LOG_INFO, "msg arrived on mailbox fd %d", retfd);
-    } else {
-        retfd = remotefd;
-        dev.log(LOG_INFO, "msg arrived on remote fd %d", retfd);
+    //it is possible both FDs have data ready concurrently.
+    if (localfd > 0 && FD_ISSET(localfd, &fds)) {
+        retfd[0] = localfd;
+        dev.log(LOG_INFO, "msg arrived on mailbox fd %d", retfd[0]);
+    }
+    if (remotefd > 0 && FD_ISSET(remotefd, &fds)) {
+        retfd[1] = remotefd;
+        dev.log(LOG_INFO, "msg arrived on remote fd %d", retfd[1]);
     }
-    return retfd;
+    return 0;
 }
 
 /*
- * Fetch sw channel msg from local mailbox fd, process it by passing it through
- * to socket fd or by the callback.
+ * Fetch sw channel msg from local mailbox fd
  */
-int processLocalMsg(pcieFunc& dev, int localfd, int remotefd, msgHandler cb)
+std::shared_ptr<sw_msg> getLocalMsg(pcieFunc& dev, int localfd)
 {
     size_t msgsz = getMailboxMsgSize(dev, localfd);
     if (msgsz == 0)
-        return -EINVAL;
+        return nullptr;
 
     std::shared_ptr<sw_msg> swmsg = std::make_shared<sw_msg>(msgsz);
     if (swmsg == nullptr)
-        return -ENOMEM;
+        return nullptr;
 
-    if (!readMsg(dev, localfd, swmsg.get()))
-        return -EINVAL;
+    if (readMsg(dev, localfd, swmsg.get()) != 0)
+        return nullptr;
 
-    int pass;
-    std::shared_ptr<sw_msg> swmsgProcessed;
-    if (!cb) {
-        // Continue passing received msg to local mailbox.
-        swmsgProcessed = swmsg;
-        pass = FOR_REMOTE;
-    } else {
-        pass = (*cb)(dev, swmsg, swmsgProcessed);
-    }
-
-    bool sent;
-    if (pass == FOR_LOCAL)
-        sent = sendMsg(dev, localfd, swmsgProcessed.get());
-    else if (pass == FOR_REMOTE)
-        sent = sendMsg(dev, remotefd, swmsgProcessed.get());
-    else // Error occured
-        return pass;
-
-    return sent ? 0 : -EINVAL;
+    return swmsg;
 }
 
 /*
  * Fetch sw channel msg from remote socket fd, process it by passing it through
  * to local mailbox fd or by the callback.
  */
-int processRemoteMsg(pcieFunc& dev, int localfd, int remotefd, msgHandler cb)
+std::shared_ptr<sw_msg> getRemoteMsg(pcieFunc& dev, int remotefd)
 {
     size_t msgsz = getSockMsgSize(dev, remotefd);
     if (msgsz == 0)
-        return -EAGAIN;
+        return nullptr;
 
     if (msgsz > 1024 * 1024 * 1024)
-        return -EMSGSIZE;
+        return nullptr;
 
     std::shared_ptr<sw_msg> swmsg = std::make_shared<sw_msg>(msgsz);
     if (swmsg == nullptr)
-        return -ENOMEM;
+        return nullptr;
 
-    if (!readMsg(dev, remotefd, swmsg.get()))
-        return -EAGAIN;
+    if (readMsg(dev, remotefd, swmsg.get()) != 0)
+        return nullptr;
 
+    return swmsg;
+}
+
+/*
+ *  passing the msg directly or the processed msg by the callback 
+ *  to local mailbox or the peer side
+ */
+int handleMsg(pcieFunc& dev, struct queue_msg &msg)
```

Comment:
`struct ` is redundant in C++ compared to C.

Reply:
will change